### DISCUSSION TITLE
Session variables are not persisted correctly

### DISCUSSION THREAD

**Discussion by egonolieux** [symfony/symfony#54856] [Q&A]
### Symfony version(s) affected

7.0.7

### Description

I have an event listener in which I write some data to the session. The data basically contains some information about the last request. However, upon the next request, the data from the session doesn't seem to correspond to the previous request, although no other/intermediate requests were made.

Here is a simplified tested example:

```php
<?php

declare(strict_types=1);

namespace App\EventListener;

use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpKernel\Event\ControllerEvent;
use Symfony\Component\HttpKernel\KernelEvents;

final class ExampleListener implements EventSubscriberInterface
{
    /**
     * @return array<string, array<int, string|int>>
     */
    public static function getSubscribedEvents(): array
    {
        return [KernelEvents::CONTROLLER => ['onKernelController', 0]];
    }

    public function onKernelController(ControllerEvent $event): void
    {
        $request = $event->getRequest();
        if (!$event->isMainRequest())
        {
            return;
        }

        $session = $request->getSession();
        $time = time();
        $previousTime = $session->get('previousTime');

        $session->set('previousTime', $time);

        dump('current: ' . $time);
        dump('previous: ' . $previousTime);
    }
}
```
**The expected output**

Notice that the previous value of request 2/3 aligns with the current value of request 1/2.

```txt
Request 1

"current: 1715050684"
"previous: 1715050677"

Request 2

"current: 1715050703"
"previous: 1715050684"

Request 2

"current: 1715050717"
"previous: 1715050703"
```

**The actual output**

Notice that the previous value of request 2/3 doesn't align with the current value of request 1/2.

```txt

Request 1

"current: 1715050684"
"previous: 1715050677"

Request 2

"current: 1715050703"
"previous: 1715050686"

Request 3

"current: 1715050717"
"previous: 1715050705"
```

---

The issue seems to be independent from the way sessions are stored. I was able to reproduce this for both file and database based sessions.

Using the native `$_SESSION` produces the expected output without issues as well, so this seems related to Symfony specifically.

### How to reproduce

Simply run the code above.

**Comment by MatTheCat** [comment]
I guess you’re using the WebProfilerBundle? Your listener would be called when it renders its toolbar, or for every asynchronous request your application would handle.
